hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java

```diff
@@ -76,6 +80,7 @@ public MapQueryEngineImpl(MapServiceContext mapServiceContext) {
         this.partitionService = nodeEngine.getPartitionService();
         this.logger = nodeEngine.getLogger(getClass());
         this.queryResultSizeLimiter = new QueryResultSizeLimiter(mapServiceContext, logger);
+        this.queryOptimizer = QueryOptimizerFactory.newOptimizer(nodeEngine.getGroupProperties());
```

Comment:
Personally I'm not crazy about such factories. Perhaps it is a flavor thing; but can't you inject the right query optimizer as constructor arg and deal with the construction when the QueryEngineImpl is made?